lib/core/src/Cardano/Api/Extra.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE Rank2Types #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+-- Module containing extra 'Cardano.Api' functionality needed by the wallet.
+module Cardano.Api.Extra
+    ( withShelleyBasedTx
+    , asAnyEraTx
+    , asAnyShelleyBasedEra
+    ) where
+
+import Prelude
+
+import Cardano.Api
+    ( CardanoEra (..)
+    , InAnyCardanoEra (..)
+    , InAnyShelleyBasedEra (..)
+    , IsCardanoEra (cardanoEra)
+    , IsShelleyBasedEra
+    , ShelleyBasedEra (..)
+    , Tx
+    )
+
+-- | Apply an era-parameterized function to an existentially wrapped
+-- tx.
+withShelleyBasedTx
+    :: InAnyShelleyBasedEra Tx
+    -> (forall era. IsShelleyBasedEra era => Tx era -> a)
+    -> a
+withShelleyBasedTx (InAnyShelleyBasedEra _era tx) f
+    = f tx
+
+-- | "Upcast" an existentially wrapped tx.
+asAnyEraTx :: forall era. IsShelleyBasedEra era => Tx era -> InAnyCardanoEra Tx
```

Comment:
Hm. Might want to tweak the wording. This is not quite the inverse of `asAnyShelleyBasedEra` because this function takes a `forall era. Tx era` rather than a `InAnyShelleyBasedEra`.